### PR TITLE
[v7r1] TransformationCleaningAgent will (re)clean very old transformations that are still in the system

### PR DESCRIPTION
**Pull request by fstagni**
BEGINRELEASENOTES

Transformation
NEW: TransformationCleaningAgent will (re)clean very old transformations that are still in the system

ENDRELEASENOTES

### CODE REVIEW

File: TransformationSystem/Agent/MCExtensionAgent.py [hunk 1]
```diff
@@ -80,7 +80,7 @@ def extendTransformation(self, transID, maxTasks):
     else:
       statusDict = res['Value']
     gLogger.verbose("Current task count for transformation %d" % transID)
-    for status in sorted(statusDict.keys()):
+    for status in sorted(list(statusDict)):
```
Comment (chaen) [inline TransformationSystem/Agent/MCExtensionAgent.py:83]: `sorted` already does a copy so you don't need `list`, do you ?

File: TransformationSystem/Agent/TransformationCleaningAgent.py [hunk 2]
```diff
@@ -124,6 +125,49 @@ def initialize(self):
     # # file catalog client
     self.metadataClient = FileCatalogClient()
 
+    # Only at (re)start: will clean ancient transformations (remnants)
+    # 1) get the transformation IDs of jobs that are older than 1 year
+    # 2) find the status of those transformations. Those "Cleaned" and "Archived" will be
+    # cleaned and archived (again)
+    res = JobMonitoringClient().getJobGroups(None, datetime.utcnow() - timedelta(days=365))
```
Comment (chaen) [inline TransformationSystem/Agent/TransformationCleaningAgent.py:132]: out of curiosity, why redoing cleaned and archived ? 
Isn't that a bit "violent" ? Should it be a script to be launched every now and then ?

Comment (fstagni) [reply TransformationSystem/Agent/TransformationCleaningAgent.py:132]: Basically, it's a race:
- the production manager submits a transformation
- TransformationAgent, and a bit later the WorkflowTaskAgent, put such transformation in their internal queue, so eventually during their (long-ish) cycle they'll work on it.
- 1 minute after creating the transformation, the production manager cleans it (by hand, for whatever reason). So, the status is changed to "Cleaning"
- the TransformationCleaningAgent cleans what has been created (maybe, nothing), sets the transformation status to "Cleaned" or "Archived"
- a bit later the TransformationAgent, and a bit later the WorkflowTaskAgent, kick in, creating tasks and jobs for a production that's effectively cleaned, but these 2 agents don't know yet. 

Of course, one could make one final check in TransformationAgent or WorkflowTaskAgent but these 2 agents are already doing a lot of stuff, and are pretty heavy. So, we should just clean from time to time. What I added here is done only when the agent (re)start, and it's quite light-ish operation anyway.

Comment (chaen) [reply TransformationSystem/Agent/TransformationCleaningAgent.py:132]: Oh I see. Wouldn't it be easier in step 5. to have these agents checking that the Transformation is still worth working on it ? (i.e. not cleaned or something else) 
This is for example something which is done in the RMS to handle canceled requests

Comment (fstagni) [reply TransformationSystem/Agent/TransformationCleaningAgent.py]: > Oh I see. Wouldn't it be easier in step 5. to have these agents checking that the Transformation is still worth working on it ? (i.e. not cleaned or something else)

No, and answer I wrote above: 
> Of course, one could make one final check in TransformationAgent or WorkflowTaskAgent but these 2 agents are already doing a lot of stuff, and are pretty heavy.

**Comment by fstagni** [comment]
@fstagni add the explanation above in the code.

File: TransformationSystem/Agent/TransformationCleaningAgent.py [hunk 3]
```diff
@@ -190,6 +193,63 @@ def execute(self):
       self.log.error("Could not get the transformations", res['Message'])
     return S_OK()
 
+  def finalize(self):
+    """ Only at finalization: will clean ancient transformations (remnants)
+        1) get the transformation IDs of jobs that are older than 1 year
```
Comment (andresailer) [inline TransformationSystem/Agent/TransformationCleaningAgent.py:198]: Should have a new line before starting the list:

https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#enumerated-lists
```suggestion

        1) get the transformation IDs of jobs that are older than 1 year
```

File: TransformationSystem/Agent/TransformationCleaningAgent.py [hunk 4]
```diff
@@ -190,6 +193,63 @@ def execute(self):
...
+        2) find the status of those transformations. Those "Cleaned" and "Archived" will be
+        cleaned and archived (again)
```
Comment (andresailer) [inline TransformationSystem/Agent/TransformationCleaningAgent.py:200]: A few characters indent to keep this as part of the list (I think) (Not sure about exact number of spaces)
```suggestion
            cleaned and archived (again)
```

**Comment by andresailer** [comment]
Is the `finalize` function also called when restarting an agent or just when it reaches its end of cycles?

**Comment by fstagni** [comment]
It's called at the end of the cycles and at re-starts and stops.

**Pull request merged**